### ISSUE TITLE
Update to sonar-java-7.0.0.26422

### ISSUE THREAD

**Issue by slarse** [SpoonLabs/sorald#331]
This is a bit tricky as they've changed some of the internals. All incompatibilities are confined within `RuleVerifier`, however, so the problems are closely contained. It should be manageable.

Currently using this issue to keep track of compatibility problems. Here are the currently known problems with different versions of sonar-java:

* 6.10+: Breaks issue filtering mechanisms, see https://github.com/SpoonLabs/sorald/issues/331#issuecomment-802678077
* 6.13.0.25138
    - Breaks rule 2755 (XxeProcessingProcessor), it just doesn't detect anything anymore
    - Gets a lot more grumpy about files that don't compile than previously (some NOCOMPILE test cases fail)

**Comment by slarse** [comment]
New version is out: https://github.com/SonarSource/sonar-java/releases/tag/6.12.0.24852

**Comment by slarse** [comment]
New version again: https://github.com/SonarSource/sonar-java/releases/tag/6.13.0.25138

I should probably get around to this upgrade soon :D

**Comment by slarse** [comment]
There are two major problems with the update:

The hack I used to be able to leverage `NOSONAR` and `@SuppressWarnings` no longer works, because that part of the internals changed.

The `XxeProcessingCheck` is no longer working at all. This may have something to do with bad classpaths. Not sure.

Overall, for this update I probably will need to revise how we leverage SonarJava. I'm trying to go as high-level as possible. I'm even considering switching over to use the SonarLint language server, but the downside of that is that we'd need to start a separate process.

**Comment by slarse** [comment]
The XxeProcessingCheck is working fine in `6.12.0.24852`. There was a change to that check in 6.13, and they probably messed it up.

For now, I'm targeting 6.12, in which the problem with `NOSONAR` and `@SuppressWarnings` still remains.

**Comment by slarse** [comment]
I've now figured out that the problems with `NOSONAR` and `@SuppressWarnings` is that the API we used before was reverted to an old API to fix some problem (see https://jira.sonarsource.com/browse/SONARJAVA-3241 and https://jira.sonarsource.com/browse/SONARJAVA-3544). Now, we don't really want to adapt to the reverted version as it itself uses deprecated APIs.

For now, we'll just stick with sonar-java 6.9 until they get rid of the deprecated APIs in the issue filtering mechanisms. Hopefully, they'll go back to the API that we're currently using, and then upgrading should be smooth sailing!

**Comment by slarse** [comment]
New major release, now it's really time to upgrade: https://github.com/SonarSource/sonar-java/releases/tag/7.0.0.26422

**Comment by slarse** [comment]
All in all, I need to completely rebuild the `RuleVerifier` class. Everything has changed.

I'll start with an effort to "move up the stack" so we don't have to rely on the internals. Last time I did that it all failed miserably as I couldn't find a way for one app to both be the producer and consumer of Sonar warnings, but perhaps I can find a way around it. If each version update is going to be this painful, it's worth putting significant effort in getting out of the internals.

**Comment by slarse** [comment]
I'm now starting to investigate if we instead can build Sorald around SonarLint, and load SonarJava as a plugin (the way it's intended). I've managed to get a minimal proof of concept working, but there are a lot of minor problems to sort out. Not least of which, it may not be possible to have SonarJava as a Maven dependency, perhaps we need to download the jarfile separately instead.

I'm looking at this approach because the upgrade to SonarJava 7+ broke some stuff which I'm frankly not sure how to fix at this point. As the update from 5 to 6 _also_ broke stuff that took me quite a while to fix, I think it would be much preferable if we step up to the public APIs of SonarLint core.

**Comment by slarse** [comment]
This is _definitely_ what we want to do. I just ran the exact same code using the SonarLint core APIs, loading first SonarJava 6.9 and then SonarJava 7.0: I had to change nothing at all.

Therefore, instead of migrating to SonarJava 7.0, the best course of action is to migrate to using the SonarLint APIs, and then using any somewhat recent version of SonarJava is pretty much no extra work.

**Comment by slarse** [comment]
And I should add, this was using only public APIs.

**Comment by algomaster99** [comment]
Hi all,

Good news! I just ran our test suites with sonar-java `v7.11.0.29148` and all our tests passed except 3 of them. I investigated the latter and none of them cause a regression. Thus, I feel we can upgrade the version of the Jar soon.

I also think we need to introduce a test case that would keep a track of what all rules are scanned by Sorald as of now. This will help us inform our users about rules that are no longer supported and the new rules introduced. We need this because the release timelines of Sorald and sonar-java are not synchronized.
